### PR TITLE
added support for textlocal sms provider

### PR DESCRIPTION
**Pull request by bhanuc**
## What kind of change does this PR introduce?

This PR adds support for textlocal.in, which is an Indian SMS provider (rates < 1/10 of twillio in India).

## Additional context

The implementation is loosely based on messagebird implementation.  Apart from signing up on the website, specific sms templates needs to be approved, only then sms can be sent,  so a list can be shared in the docs for the same. 

Note: template approval is a mandate of the indian government and all companies have to follow it for sending SMS in India.

### CODE REVIEW

File: conf/configuration.go [hunk 1]
```diff
@@ -365,3 +371,13 @@ func (t *MessagebirdProviderConfiguration) Validate() error {
 	}
 	return nil
 }
+
+func (t *TextLocalProviderConfiguration) Validate() error {
+	if t.ApiKey == "" {
+		return errors.New("Missing Messagebird access key")
```
Comment (kangmingtay) [inline conf/configuration.go:377]: this should be textlocal instead of messagebird?

**Review by kangmingtay — CHANGES_REQUESTED** [review]
hi @bhanuc, thanks for your contribution! Can you please run through the guidelines here for adding a new sms provider (https://github.com/supabase/gotrue/blob/master/CONTRIBUTING.md#guidelines-for-implementing-additional-oauth-providers). It's similar to how the oauth2 providers are implemented so an end-to-end test is required.

File: conf/configuration.go (hunk 1 above)
Comment (bhanuc) [reply conf/configuration.go]: updated - btw I am new to golang, so not very confident about the test cases, I tried looking for twillio and messagebird, test cases but could not find any. can you share the same if possible.

Comment (kangmingtay) [reply conf/configuration.go]: sure, we currently don't have an end-to-end test setup in gotrue for the sms providers but you can test it manually by doing the following steps:
1. set the correct env vars for enabling the textlocal provider
2. start gotrue locally 
3. hit the /otp endpoint and check that you receive an sms from textlocal 
4. hit the /verify endpoint with the otp received and check that the user is verified

**Comment by kangmingtay** [comment]
Hi @bhanuc, did you try testing out the textlocal sms provider by doing [the following](https://github.com/supabase/gotrue/pull/305#discussion_r777812074) yet? I don't think i'm able to signup for a textlocal account since it's not available in my country.

**Comment by devkiran** [comment]
@kangmingtay I tested this implementation locally, it didn't work.

Got the following messaga
```
{
    "code": 400,
    "msg": "Error sending sms otp: Sms Provider textlocal could not be found"
}
```

We need to add a new case in `sms_provider.go` for textlocal. I have an API from textlocal.

**Comment by kangmingtay** [comment]
hey @devkiran, this is amazing, thanks for helping out @bhanuc with this PR :)

**Comment by bhanuc** [comment]
have added changes suggested and also tested it, though I see devkiran has also made the same pr, I am ok with his changes getting merged, as he did provide with good inputs and incorporated post request over GET.

**Comment by devkiran** [comment]
Thanks @bhanuc for the contribution.

**Comment by bharath2612** [comment]
Hey @bhanuc and @devkiran we are getting this error on sending otps with textlocal, the sender and template are approved from textlocal's end. Can you guide us here? 

This was the error : AuthException(message: Error sending confirmation sms: textlocal error: Internal Error, statusCode: 400)

**Comment by kangmingtay** [comment]
hey @bharath2612, please take a look at the comment here: https://github.com/supabase/gotrue/pull/342#issuecomment-1445632340

**Pull request closed without merging**